pyads/symbol.py

```diff
@@ -0,0 +1,271 @@
+
+        See Connection.add_device_notification(...).
+
+        When `attr` is omitted, the default will be used.
+
+        The notification handles are returned but also stored locally. When
+        this symbol is destructed any notifications will be freed up
+        automatically.
+        """
+
+        if attr is None:
+            attr = NotificationAttrib(length=sizeof(self.plc_type))
+
+        handles = self._plc.add_device_notification(
+            (self.index_group, self.index_offset), attr, callback, user_handle
+        )
+
+        self._handles_list.append(handles)
+
+        return handles
+
+    def clear_device_notifications(self):
+        """Remove all registered notifications"""
+        if self._handles_list:
+            for handles in self._handles_list:
+                self._plc.del_device_notification(*handles)
+            self._handles_list = []  # Clear the list
+
+    def del_device_notification(self, handles: Tuple[int, int]):
+        """Remove a single device notification by handles"""
+        if handles in self._handles_list:
+            self._plc.del_device_notification(*handles)
+            self._handles_list.remove(handles)
+
+    @staticmethod
+    def get_type_from_str(type_str: str) -> Any:
+        """Get PLCTYPE_* from PLC name string
+
+        If PLC name could not be mapped, return None. This is done on
+        purpose to prevent a program from crashing when an unusable symbol
+        is found. Instead, exceptions will be thrown when this unmapped
+        symbol is read/written.
+        """
+
+        # If simple scalar
+        plc_name = "PLCTYPE_" + type_str
+        if hasattr(constants, plc_name):
+            # Map e.g. 'LREAL' to 'PLCTYPE_LREAL' directly based on the name
+            return getattr(constants, plc_name)
+
+        # If ARRAY
+        reg_match = re.match(r"ARRAY \[(\d+)..(\d+)\] OF (.*)", type_str)
+        if reg_match is not None:
+
+            groups = reg_match.groups()
+            size = int(groups[1]) + 1 - int(groups[0])  # Estimate the size
+            scalar_type_str = groups[2]
+
+            # Find scalar type
+            scalar_type = AdsSymbol.get_type_from_str(scalar_type_str)
+
+            if scalar_type:
+                return scalar_type * size
+
+            # Fall to method default instead
+
+        # If array/matrix (an 1D array is also called a matrix)
+        reg_match = re.match(r"matrix_(\d+)_(.*)_T", type_str)
+        if reg_match is not None:
+
+            groups = reg_match.groups()
+            size = int(groups[0])
+            scalar_type_str = groups[1]
+
+            if scalar_type_str in constants.PLC_ARRAY_MAP:
+                return constants.PLC_ARRAY_MAP[scalar_type_str](size)
+
+        # We allow unmapped types at this point - Instead we will throw  an
+        # error when they are being addressed
+
+        return None
+
+    @staticmethod
+    def get_type_from_int(type_int: int) -> Any:
```

Comment:
Is this still needed? As I recall we now use only strings to get the PLCTYPE, right?

Reply:
You're right, I'll delete it.